specs/sharding/polynomial-commitments.md

```diff
@@ -0,0 +1,387 @@
+    return [BLSFieldElement(int.from_bytes(x, "little")) for x in sliced_block]
+```
+
+## Polynomial operations
+
+#### `add_polynomials`
+
+```python
+def add_polynomials(a: BLSPolynomialByCoefficients, b: BLSPolynomialByCoefficients) -> BLSPolynomialByCoefficients:
+    """
+    Sums the polynomials `a` and `b` given by their coefficients
+    """
+    a, b = (a, b) if len(a >= b) else (b, a)
+    return [(a[i] + (b[i] if i < len(b) else 0)) % BLS_MODULUS for i in range(len(a))]
+```
+
+#### `multiply_polynomials`
+
+```python
+def multiply_polynomials(a: BLSPolynomialByCoefficients, b: BLSPolynomialByCoefficients) -> BLSPolynomialByCoefficients:
+    """
+    Multiplies the polynomials `a` and `b` given by their coefficients
+    """
+    r = [0]
+    for power, coef in enumerate(a):
+        summand = [0] * power + [coef * x % BLS_MODULUS for x in b]
+        r = add_polynomials(r, summand)
+    return r
+```
+
+
+#### `interpolate_polynomial`
+
+```python
+def interpolate_polynomial(xs: List[BLSFieldElement], ys: List[BLSFieldElement]) -> BLSPolynomialByCoefficients:
+    """
+    Lagrange interpolation
+    """
+    assert len(xs) == len(ys)
+    r = [0]
+
+    for i in range(len(xs)):
+        summand = [ys[i]]
+        for j in range(len(ys)):
+            if j != i:
+                weight_adjustment = bls_modular_inverse(xs[j] - xs[i])
+                summand  = multiply_polynomials(summand, [weight_adjustment, ((MODULUS - weight_adjustment) * xs[i])])
+        r = add_polynomials(r, summand)
+    
+    return r
+```
+
+#### `evaluate_polynomial_in_evaluation_form`
+
+```python
+def evaluate_polynomial_in_evaluation_form(poly: BLSPolynomialByEvaluations, x: BLSFieldElement) -> BLSFieldElement:
+    """
+    Evaluates a polynomial (in evaluation form) at an arbitrary point
+    """
+    field_elements_per_blob = SAMPLES_PER_BLOB * FIELD_ELEMENTS_PER_SAMPLE
+    roots = roots_of_unity(field_elements_per_blob)
+
+    def A(z):
+        r = 1
+        for w in roots:
+            r = r * (z - w) % BLS_MODULUS
+        return r
+
+    def Aprime(z):
+        return field_elements_per_blob * pow(z, field_elements_per_blob - 1, BLS_MODULUS) 
+
+    r = 0
+    inverses = [bls_modular_inverse(z - x) for z in roots]
+    for i, x in enumerate(inverses):
+        r += poly[i] * bls_modular_inverse(Aprime(roots[i])) * x % self.BLS_MODULUS
+    r = r * A(x) % self.BLS_MODULUS
+    return r
+```
+
+## KZG Operations
+
+We are using the KZG10 polynomial commitment scheme (Kate, Zaverucha and Goldberg, 2010: https://www.iacr.org/archive/asiacrypt2010/6477178/6477178.pdf).  
+
+### Elliptic curve helper functoins
+
+#### `elliptic_curve_lincomb`
+
+```python
+def elliptic_curve_lincomb(points: List[KZGCommitment], scalars: List[BLSFieldElement]) -> KZGCommitment:
+    """
+    BLS multiscalar multiplication. This function can be optimized using Pippenger's algorithm and variants. This is a non-optimized implementation.
+    """
+    r = bls.Z1()
```

Comment:
I assume `Z1` is the point at infinity in G1, but I didn't find a reference to this somewhere.